dashboard/app/models/lesson.rb

```diff
@@ -730,13 +730,12 @@ def lesson_plan_has_verified_resources
   # - be migrated
   # - be in a course version
   # - be in course versions from the same version year
-  def copy_to_script(destination_script)
+  def copy_to_script(destination_script, new_level_suffix = nil)
     return if script == destination_script
     raise 'Both lesson and script must be migrated' unless script.is_migrated? && destination_script.is_migrated?
     raise 'Destination script and lesson must be in a course version' if destination_script.get_course_version.nil? || script.get_course_version.nil?
-    raise 'Destination script must have the same version year as the lesson' unless destination_script.get_course_version.version_year == script.get_course_version.version_year
 
-    ActiveRecord::Base.transaction do
+    ActiveRecord::Base.transaction(requires_new: true, joinable: false) do
```

Comment:
This is the first time I've seen these options. It looks like they are related to preventing problems when transactions are nested. Cool! Can you share what you found about when these params should be used? I see some varying opinions.

Other questions/comments on transactions:
1. could we remove this transaction, and put one in LessonsController#clone instead? that could help avoid nested transactions completely.
2. transactions will rollback db contents, but levelbuilder could have updated corresponding level files, which won't get rolled back. I'm not sure this is worth handling in your code, but could be worth looking out for if you ever do hit an error.

Reply:
I read about `require_new` [here](https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html#module-ActiveRecord::Transactions::ClassMethods-label-Nested+transactions). `joinable` is significantly less well-documented though. Pretty much everything I find leads back to [a blog](https://makandracards.com/makandra/42885-nested-activerecord-transaction-pitfalls). It's also referenced, though not explained, on [api dock](https://apidock.com/rails/ActiveRecord/ConnectionAdapters/DatabaseStatements/transaction). I did test that this setup did what I wanted (rollback the entire script copy if a lesson fails to copy).

> could we remove this transaction, and put one in LessonsController#clone instead? that could help avoid nested transactions completely.

I guess it depends what we want out of these transactions. I added them for confidence that even as a developer running a command in dashboard-console, the DB never gets in a bad state. However, cleaning up a lesson copy is pretty easy and we could keep the transaction on script copy and add one in the lesson controller like you suggested. That option (plus removing the transaction from lesson group) is growing on me...

> transactions will rollback db contents, but levelbuilder could have updated corresponding level files, which won't get rolled back

Yeah, I'm not sure what to do here. I could try to delete the level files on rollback but I'm not sure if it's worth it. Definitely open to other thoughts here

Reply:
1. from the blog post, my understanding is that `joinable` makes a transaction safe to have other transactions inside it, and `require_new` makes it safe for a transaction to have other transactions outside it. So my takeaway is that (if the blog post is right) then adding a transaction with both of these flags is always safe from the nested transaction problem. Is that your understanding as well? If we feel confident about this, then we could embrace nested transactions and not do extra work to avoid them in the future.

2. yeah, I don't see a code improvement here that's simple enough to be worth doing. 

the tricky thing about levelbuilder is that someone could have added a new level right as you added 100s of levels, so it is hard to go back and remove all the level files you just added without risking removing a curriculum writer's work. 

when we are doing yearly copies, we can probably use a heuristic on the level filename. For example, when we do this in January, most level files will be gaining the 2022 suffix, and most newly created levels will not have that suffix (though I'm not 100% sure on that).

heading into meetings now, will finish reviewing this afternoon.